Add lark sheets connector

7c00 opened the pull request:
This pull introduces Lark Sheets Connector, which is inspired by [Google Sheets Connector](https://github.com/prestodb/presto/pull/16721). Though there are tiny differences between them. 

Developers could run `LarkSheetsQueryRunners#main`, and then follow the _QueryModel_ section from the [doc](https://github.com/7c00/presto/blob/presto-lark/presto-docs/src/main/sphinx/connector/larksheets.rst) to try the connector.

Test plan

- Integation test
- Unit test


```
== RELEASE NOTES ==

Connectors
* Introduce Lark sheets connector.

```

7c00 commented:
Hello, @ChunxuTang @beinan. Could you help review this PR?

rohanpednekar commented:
Thanks, @7c00 for your contribution, we were planning to request @ChunxuTang and @beinan as they helped us earlier to review the google sheet connector.

ChunxuTang commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsConfig.java:
```diff
@@ -0,0 +1,64 @@
+        FEISHU, LARK;
+    }
+
+    private Domain appDomain = Domain.FEISHU;
```
Just wondering what's the difference between `FEISHU` and `LARK`? This is the lark connector, but the default domain is `FEISHU`?

7c00 replied:
Lark and Feishu are the same product but serving different users. Lark serves global users, while Feishu serves Chinese users. They have different SDK domains.

I have changed the default value to LARK.

ChunxuTang commented on presto-lark-sheets/src/test/resources/testing-app-secret-base64:
```diff
@@ -0,0 +1 @@
+ewogICJhcHAtc2VjcmV0IiA6ICJYSHdrckVna0N6Y3dFZ0NDNlpQQjRnWXZtb25OSmdNTiIKfQ==
```
Is there a TTL for this secret? And is there an owner email account? There's a potential risk here that when the secret expires, it may break presto CI workflows.

7c00 replied:
There is no TTL for the secret. The secret is not bound to an email account. 
I wonder if we should disable `TestSheetsIntegration` by default as `TestGoogleSheets`.

ChunxuTang replied:
Yeah, I think we may need to disable it.
Hi @beinan, could you help disable this integration test by default?

beinan replied:
Yes, let's disable the test, otherwise it would be too flaky I guess

We could use ```@Test(enabled = false)```  in the integration test.

7c00 replied:
Disabled all the tests in `TestLarkSheetsIntegration`.

ChunxuTang reviewed (commented):
Hi @7c00, thanks a lot for your work! This looks to be a very nice feature, and the Presto community could benefit.
Just left some suggestions/comments. For the first round of review, I mainly focused on the coding styles. Thanks again for your contributions!

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:
```diff
@@ -0,0 +1,342 @@
+    public List<ConnectorTableLayoutResult> getTableLayouts(ConnectorSession session,
+            ConnectorTableHandle table,
+            Constraint<ColumnHandle> constraint,
+            Optional<Set<ColumnHandle>> desiredColumns)
```
The parameter `desiredColumns` is definied by the SPI method [ConnectorMetadata#getTableLayouts](https://github.com/prestodb/presto/blob/d37ac0735773af0cd17b26fff1e2c9a9575734d8/presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java#L113). So we could not remove it.

ChunxuTang replied:
Thanks for the clarification!

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:
```diff
@@ -0,0 +1,342 @@
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        return Streams.stream(schemaStore.listForUser(session.getUser()))
```
Yes. I have replaced Streams.stream with StreamSupport.stream.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:
```diff
@@ -0,0 +1,342 @@
+        else {
+            List<SheetInfo> filtered = filterSheets(metaInfo.getSheets(), prefixTableName);
+            if (filtered.size() == 1) {
+                // a trick here
```
Added some comments to the code for details. 
>    // Use prefixTableName (other than SheetInfo#getTitle) to create SchemaTableName
>    // in order to make queries like `DESC "@sheetId"` or `DESC "#1"` work

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsRecordCursor.java:
```diff
@@ -0,0 +1,224 @@
+public class SheetsRecordCursor
+        implements RecordCursor
+{
+    private final ConnectorSession session;
```
Removed.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsSplit.java:
```diff
@@ -0,0 +1,66 @@
+    {
+        return ImmutableMap.builder()
+                .put("table", table)
+                .build();
```
Fixed.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java:
```diff
@@ -0,0 +1,249 @@
+                return new SpreadsheetInfo(data.getSpreadsheetToken(), properties.getTitle(), properties.getRevision(), sheets);
+            }
+            catch (Exception e) {
+                throw wrapApiError(e, "Could not to get metadata of spreadsheet " + token);
```
Fixed.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsRecordCursor.java:
```diff
@@ -0,0 +1,224 @@
+        return sum;
+    }
+
+    private static int[] createIndexMapping(List<SheetsColumnHandle> readColumns)
```
The static method is placed here since its logic is not general but very sepecial to SheetsRecordCursor.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java:
```diff
@@ -0,0 +1,249 @@
+                sheet.getRowCount());
+    }
+
+    private static Domain toLarkDomain(SheetsConfig.Domain domain)
```
Moved the utility function like `loadAppSecret` to SheetsUtil.

For `toLarkDomain`, it translates SheetsConfig.Domain to Lark SDK Domain. As we encapsulated all SDK invocations into `SheetsApiFactory`, it would be better to keep `toLarkDomain` remain in this class. What do you think?

ChunxuTang replied:
Sure, thanks for the clarification.

7c00 commented:
@ChunxuTang Thanks a lot for your comments. I haved update the code. Could you please review it again?

ChunxuTang commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:
```diff
@@ -0,0 +1,343 @@
+                // Columns without name are ignored
+                continue;
+            }
+            String columnName = rawColumnName.toLowerCase(ENGLISH);
```
What if the column name is non-English? Will it throw an exception or just skip the column silently?

7c00 replied:
Presto supports UTF-8 identifier names. So for a non-english string, it can be used as a column name in Lark connector, but should be quoted as `"XXX"`to access the column.

ChunxuTang replied:
Gotcha. Thanks!

ChunxuTang commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java:
```diff
@@ -0,0 +1,225 @@
+        return new SimpleSheetsApi();
+    }
+
+    public class SimpleSheetsApi
+            implements SheetsApi
```
Just wondering as the `CachingSheetsApi` is implemented in a separate file, why do you prefer to implement the `SimpleSheetsApi` in the factory class? Any specific reasons?

7c00 replied:
Possible some of my perference during development I guess. Have rafactored SimpleSheetsApi to a top level class to make it matched the code style of CachingSheetsApi.

ChunxuTang reviewed (commented):
Hi @7c00, thanks for your revision!
Left a few minor suggestions/comments.

7c00 commented on presto-docs/src/main/sphinx/connector/larksheets.rst:
```diff
@@ -0,0 +1,105 @@
+
+    connector.name=lark-sheets
+    app-domain=FEISHU
+    app-id=cli_a16fc7ddf4b9900d
```
Yes, it is a real app id.

ChunxuTang replied:
Just wondering as this is a real ID, is it OK to expose it in an example file? Maybe just use `exampleAppId`?
Your call.

7c00 replied:
Updated to a fake app id as `example_app_id`.

7c00 commented on presto-docs/src/main/sphinx/connector/larksheets.rst:
```diff
@@ -0,0 +1,105 @@
+``app-domain``                      Use ``LARK`` if created at `Lark Open Platform`, or ``FEISHU`` otherwise
+``app-id``                          Your app id, available at `Lark Open Platform` or `Feishu Open Platform`
+``app-secret-file``                 Path to the JSON file that contains your app secret
+=================================== =====================================================================
```
Fixed.

7c00 commented on presto-docs/src/main/sphinx/connector/larksheets.rst:
```diff
@@ -0,0 +1,105 @@
+
+The ``token`` property is taken from the last path segment of the spreadsheet url, e.g.
+``shtcnBf5pg4BNSkwV2Ku5xwW9Pf`` for
+``https://test-ch80md45anra.feishu.cn/sheets/shtcnBf5pg4BNSkwV2Ku5xwW9Pf?sheet=MT1p4I``
```
No, it will not expire.

beinan replied:
To make it more readable, I might use a pattern string rather than a real link.  e.g.  https://{account_id}.feishu.ch/sheet/{token}/sheet={sheet_id}

7c00 commented on presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsPlugin.java:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    public void createTestingAppSecretFile()
```
Removed as it's not used any more.

7c00 commented on presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsUtil.java:
```diff
@@ -0,0 +1,70 @@
+/*
```
Sure. Added tests for new functions here.

ChunxuTang replied:
Did you add them? Forgot to commit changes?
It doesn't show any changes here.

7c00 replied:
The tests was added to another class by mistake. Have corrected this and moved the test to `TestLarkSheetsUtil`.

7c00 commented:
@ChunxuTang Thanks for your comments. :heart: 
Have updated the code. Could you please take a review again? Thanks in advance.

ChunxuTang commented:
> @ChunxuTang Thanks for your comments. ❤️ Have updated the code. Could you please take a review again? Thanks in advance.

Sure! I'll take another round of review very soon this week.

ChunxuTang reviewed (commented):
@7c00 Thanks for your contributions! 

As you created this PR some time ago and it is now left a bit behind the master branch, could you merge master into this feature branch to avoid conflicts?

beinan commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsConnector.java:
```diff
@@ -0,0 +1,119 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class SheetsConnector
```
To be more consistent, can we use LarkSheetsConnector as the class name?  Similarly, other class names should also use LarkSheets

7c00 commented:
@ChunxuTang @beinan Thanks for your comments. I have rebased my branch with master and updated the code as the comments. Could you please review it again? Thanks in advance.

beinan commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java:
```diff
@@ -0,0 +1,343 @@
+        LinkedHashMap<String, LarkSheetsColumnHandle> columns = new LinkedHashMap<>(numColumns);
+        for (int i = 0; i < numColumns; i++) {
+            String rawColumnName = header.get(i);
+            if (rawColumnName == null) {
```
I'm just thinking if we could use the column index when the column name is null.   Your call.

7c00 replied:
The column name can be a sequence of any UTF-8 characters. So it might be hard to define the column name format. For example, using `$2` as the column index to reference the second column, it might conflict with another column with the name of `$2`. In the other hand, updating a spreadsheet to fill a column name is very easy for end users. So column index name is not be supported.

beinan replied:
gotcha, thanks for the clarification!

beinan reviewed (commented):
just left a couple comments for some minor issues,  I will continue review this giant pr in this week.  Many thanks for this create contribution!

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsConnector.java:
```diff
@@ -0,0 +1,119 @@
+        LarkSheetsTransactionHandle transaction = new LarkSheetsTransactionHandle();
+
+        LarkSheetsApi api = apiFactory.get();
+        LarkSheetsApi cachingApi = new CachingLarkSheetsApi(api);
```
Fixed.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsErrorCode.java:
```diff
@@ -0,0 +1,48 @@
+    SCHEMA_TOKEN_NOT_PROVIDED(0x03, USER_ERROR),
+    SCHEMA_NOT_EXISTS(0x04, USER_ERROR),
+    SHEET_NAME_AMBIGUOUS(0x04, USER_ERROR),
+    SHEETS_API_ERROR(0x10, EXTERNAL),
```
Maybe a typo. Have re-organized the error codes to threw catagories.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsErrorCode.java:
```diff
@@ -0,0 +1,48 @@
+    SCHEMA_ALREADY_EXISTS(0x03, USER_ERROR),
+    SCHEMA_TOKEN_NOT_PROVIDED(0x03, USER_ERROR),
+    SCHEMA_NOT_EXISTS(0x04, USER_ERROR),
+    SHEET_NAME_AMBIGUOUS(0x04, USER_ERROR),
```
Maybe a typo. Have re-orderred the error codes to three catagories.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java:
```diff
@@ -0,0 +1,343 @@
+        }
+        else if (matched.size() > 1) {
+            throw new PrestoException(LarkSheetsErrorCode.SHEET_NAME_AMBIGUOUS,
+                    format("Ambiguous name %s in spreadsheet %s: matched sheets: %s", sheetName, token, matched));
```
Putting the raw tokens in exception messages might cause data leak. Have masked the token before writing the token.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java:
```diff
@@ -0,0 +1,343 @@
+    public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> schemaName)
+    {
+        if (!schemaName.isPresent()) {
+            throw new PrestoException(NOT_PERMITTED, "Schema is required to list tables");
```
Default schema is not supported, since we dont know which spreadsheet shoud be associated with the schema.

7c00 commented:
@ChunxuTang @beinan Hi, thanks for your review. :heart:  Do we need more review before merge this pr?

beinan reviewed (commented):
I think it looks good to me except two very minor issues.

7c00 commented on presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsRecordCursor.java:
```diff
@@ -0,0 +1,220 @@
+        this.api = requireNonNull(api, "api is null");
+        this.split = requireNonNull(split, "split is null");
+        this.readColumns = requireNonNull(readColumns, "columns is null");
+        this.indexMapping = createIndexMapping(readColumns);
```
Fixed.

7c00 commented on presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestLarkSheetsIntegration.java:
```diff
@@ -0,0 +1,246 @@
+    // Used for fast switch to enable or disable tests
+    private static final boolean TEST_ENABLED = false;
+    private static final String CATALOG = "larksheets";
+    // https://test-ch80md45anra.feishu.cn/sheets/shtcnBf5pg4BNSkwV2Ku5xwW9Pf?sheet=MT1p4I
```
This url points to the spreadsheet used for this integration test. The `TESTING_TOKEN` is extracted from the url. I have added comments to make it more clear.

7c00 commented:
ping @beinan @ChunxuTang

ChunxuTang reviewed (approved):
LGTM. @7c00 Thanks for your nice work!

7c00 commented:
Rebased with master branch.

beinan reviewed (approved):
lgtm, rerun the test

7c00 commented:
Thanks @beinan ❤️ . This CI finished. Could we merge this PR?

7c00 commented:
Thanks @beinan @ChunxuTang for your time to review!

rohanpednekar commented:
@7c00 Do you think you can provide short documentation on how to use this connector that we can publish at  https://prestodb.io/docs/current/connector.html ?

7c00 commented:
@rohanpednekar Yes. The connector doc is included in the PR. It's https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector/larksheets.rst. It's a bit strange that it's not displayed in prestodb website.

rohanpednekar commented:
Thanks @7c00 for the confirmation, I guess this ll get published with the 0.272 release notes PR
